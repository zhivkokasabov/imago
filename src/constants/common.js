export const debounceTime = 500;