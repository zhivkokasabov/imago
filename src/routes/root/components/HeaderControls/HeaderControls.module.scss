.header-controls {
  display: flex;
  align-items: center;
  gap: 2rem;

  &__search-container {
    flex-grow: 1;
    display: flex;
    gap: 2rem;
    max-width: 35rem;
    margin-right: 2rem;
    position: relative;

    button {
      position: absolute;
      left: 0;
      top: 0;
      padding: 0.75rem 1rem;
      color: rgb(var(--coolgrey-50));
      z-index: 1;
    }
  }

  &__input input {
    height: 2.5rem;
    padding: 0 0 0 3rem;
    border: 1px solid rgb(var(--coolgrey-50));

    &::placeholder {
      color: rgb(var(--coolgrey-50));
      font-size: .875rem;
    }
  }
}