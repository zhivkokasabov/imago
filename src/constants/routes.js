import { queryKeys } from './common';

export const routes = {
  home: '/',
  editorial: '/',
  sports: '/',
  creative: '/',
  archive: '/',
  account: '/',
  search: (searchTerm) => searchTerm ? `/search?${new URLSearchParams({ [queryKeys.searchTerm]: searchTerm })}` : '/search',
  product: (id) => `/products/${id}`
};
