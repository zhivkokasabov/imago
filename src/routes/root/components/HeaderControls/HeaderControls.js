import React, { useEffect, useState } from 'react';
import { Link, useNavigate, useLocation } from 'react-router-dom';

import Image from '../../../../components/Image/Image';
import SearchField from '../../../../components/SearchField/SearchField';
import Button from '../../../../components/Button/Button';
import SearchIcon from '../../../../icons/SearchIcon';

import styles from './HeaderControls.module.scss';

import { routes } from '../../../../constants/routes';

import classNames from 'classnames';

export default function HeaderControls({ className }) {
  const [searchTerm, setSearchTerm] = useState();
  const [showSearchBar, setShowSearchBar] = useState(true);
  let navigate = useNavigate();
  let { pathname } = useLocation();

  const onSearchChange = ({ target: { value } }) => setSearchTerm(value);
  const onKeyDown = ({ key }) => key === 'Enter' && navigate(routes.search(searchTerm));
  const onSearchIconClick = () => navigate(routes.search(searchTerm));

  useEffect(() => {
    if (pathname === routes.search()) {
      setShowSearchBar(false);
    } else {
      setShowSearchBar(true);
    }
  }, [pathname]);

  return (
    <div className={classNames(styles['header-controls'], className)}>
      <Link to={routes.home}>
        <Image src="https://cdn.imago-images.com/Images/Logo/IMAGO-Primary_Logos-RGB-BLACK.svg" loading="eager" width={80} />
      </Link>
      {
        showSearchBar
          ? (
            <div className={styles['header-controls__search-container']}>
              <Button onClick={onSearchIconClick} type="button">
                <SearchIcon width={16} height={16} />
              </Button>
              <SearchField
                className={styles['header-controls__input']}
                placeholder="Search..."
                onChange={onSearchChange}
                onKeyDown={onKeyDown}
              />
            </div>
          )
          : (null)
      }
    </div>
  );
}