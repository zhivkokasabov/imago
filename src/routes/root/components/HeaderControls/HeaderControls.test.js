import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import HeaderControls from './HeaderControls';

import { routes } from '../../../../constants/routes';

import * as reactRouterDom from 'react-router-dom';

jest.mock('react-router-dom', () => {
  const actual = jest.requireActual('react-router-dom');

  return {
    ...actual,
    useNavigate: jest.fn(),
    useLocation: jest.fn()
  };
});

describe('HeaderControls component', () => {
  const navigateMock = jest.fn();
  const searchTerm = 'searchTerm';

  const renderComponent = (args) => render(
    <MemoryRouter {...args}>
      <HeaderControls type="button" />
    </MemoryRouter>
  );

  beforeEach(() => {
    jest.spyOn(reactRouterDom, 'useNavigate').mockReturnValue(navigateMock);
    jest.spyOn(reactRouterDom, 'useLocation').mockReturnValue({});
  });

  afterEach(jest.clearAllMocks);

  it(`should render search if pathname does not equals ${routes.search()}`, () => {
    const { asFragment } = renderComponent();

    expect(asFragment()).toMatchSnapshot();
  });

  it(`should not render search if pathname equals ${routes.search()}`, () => {
    jest.spyOn(reactRouterDom, 'useLocation').mockReturnValue({
      pathname: routes.search()
    });

    const { asFragment } = renderComponent();

    expect(asFragment()).toMatchSnapshot();
  });

  it('should navigate on enter key down', () => {
    const { getByTestId } = renderComponent();

    fireEvent.change(getByTestId('searchField'), { target: { value: searchTerm } });
    fireEvent.keyDown(getByTestId('searchField'), { key: 'Enter' });

    expect(navigateMock).toHaveBeenCalledTimes(1);
    expect(navigateMock).toHaveBeenCalledWith(routes.search(searchTerm));
  });

  it('should navigate on searchButton click', () => {
    const { getByTestId } = renderComponent();

    fireEvent.change(getByTestId('searchField'), { target: { value: searchTerm } });
    fireEvent.click(getByTestId('searchButton'));

    expect(navigateMock).toHaveBeenCalledTimes(1);
    expect(navigateMock).toHaveBeenCalledWith(routes.search(searchTerm));
  });
});
