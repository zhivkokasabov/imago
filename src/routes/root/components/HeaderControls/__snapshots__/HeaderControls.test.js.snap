// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderControls component should not render search if pathname equals /search 1`] = `
<DocumentFragment>
  <div
    class="header-controls"
  >
    <a
      href="/"
    >
      <img
        alt="logo"
        class="image"
        loading="eager"
        src="Images/Logo/IMAGO-Primary_Logos-RGB-BLACK.svg"
        width="80"
      />
    </a>
  </div>
</DocumentFragment>
`;

exports[`HeaderControls component should render search if pathname does not equals /search 1`] = `
<DocumentFragment>
  <div
    class="header-controls"
  >
    <a
      href="/"
    >
      <img
        alt="logo"
        class="image"
        loading="eager"
        src="Images/Logo/IMAGO-Primary_Logos-RGB-BLACK.svg"
        width="80"
      />
    </a>
    <div
      class="header-controls__search-container"
    >
      <button
        class="button"
        data-testid="searchButton"
        type="button"
      >
        <svg
          fill="currentColor"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z"
          />
        </svg>
      </button>
      <div
        class="search-field header-controls__input"
      >
        <input
          class="search-field__input"
          data-testid="searchField"
          placeholder="Search..."
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
